Compiler error running tests

For a few weeks now, the daily CI builds fail on nightly. Everything works up till the doctest which fails with:

```
error: extern location for quote does not exist: /home/travis/build/auto-impl-rs/auto_impl/target/debug/deps/libquote-db9ca8cbcca99a73.rlib
error[E0463]: can't find crate for `quote`
```

I investigated a bit and found the cause of this failure:
- `cargo build` (and `cargo test --no-run`) prepare the files correctly. They create `libquote-db9ca8cbcca99a73.rlib` in the right location.
- `cargo test --tests` now runs our own test harnesses. They need to get the path to `libauto_impl****.so`. This is done via the `build_plan` crate and by executing `cargo build -Z unstable-options --build-plan`. Executing this command deletes the `libquote-db9ca8cbcca99a73.rlib` file! (And apparently a bunch of other files as well)
- `cargo test --doc` then tries to run `rustdoc` which is unable to find the `libquote-db9ca8cbcca99a73.rlib` file.

So the problem is `cargo build -Z unstable-options --build-plan` removing the rlib file. I'm not yet sure if that's intended or not. When I first created those tests, it worked fine. But I guess it was clear that the current solution is hacky and would break at some point. 

I'm trying to fix this, but wanted to create this issue already to write down my findings.

---

Apparently this is tracked here: https://github.com/rust-lang/cargo/issues/6954

--- a/tests/util/mod.rs
+++ b/tests/util/mod.rs
@@ -46,6 +46,21 @@ pub(crate) fn get_dep_path() -> PathBuf {
 
     assert!(outputs.len() == 1);
 
+    // Run `cargo build` again to regenerate artifacts that might have been
+    // removed by the previous execution. See
+    // https://github.com/rust-lang/cargo/issues/6954
+    let mut command = Command::new(env!("CARGO"));
+    command.args(&["build"]);
+    command.stderr(Stdio::inherit());
+
+    #[cfg(feature = "nightly")]
+    command.arg("--features=nightly");
+
+    let status = command.status().expect("failed to run `cargo build`");
+    if !status.success() {
+        panic!("error running 'cargo build'");
+    }
+
     outputs.remove(0).into()
 }
 

